Qr login welcome

nayooti opened the pull request:
fixes #610 
fixes #608 

Realised that our QRReaderController had some bugs, so fixed that. 
I also changed that the QRReaderController is presented on top of the QR-tab (instead of being pushed).  It is presented on WelcomeScreen also, and it is a lot easier to reuse, if on each case the reused VC is either pushed or presented. And presenting makes a lot more sense too.

Whats need to be tested:
scan non valid qr-code (use a random non valid qr-code) -> should show Error message over qrReader (translations are still missing)
scan valid qr-code ->   should show Confirm-Alert (cancel, ok). 

Cancel will currently redirect to WelcomeView (is this wanted?). 
Ok, will redirect to WelcomeController (showing Spinner) while trying to create account. 

If creation fails, Error message is shown (**need some additional text here**). On success user will be redirected to Chat-Tab. (On next iteration we will add options to add Username, Image etc.). 

To test this PR you need a QR-Code (can be provided by @r10s) .

r10s commented on deltachat-ios/DC/Wrapper.swift:
```diff
@@ -139,6 +139,16 @@ class DcContext {
+        }
+        let success = dc_set_config_from_qr(contextPointer, qrCode)
+        configure()
+        return success == 1
```
i think, this is wrong. configure() always returns immediately and error/success is reported by DC_EVENT_CONFIGURE_PROGRESS, see AccountSetupController to get an idea, maybe copy/paste the code from there ;)

nayooti replied:
this has been fixed.

nayooti commented:
I added Progress Alert (AccountSetupController and WelcomeViewController use now the same ProgressAlert-code).

In combination with progress alert I wasn't happy with the spinner, so I changed that, too. Spinner only replaces the buttons on the welcome view  now.

Hope you like it.

cyBerta commented on deltachat-ios/Controller/QrViewController.swift:
```diff
@@ -318,3 +240,81 @@ class QrViewController: UITableViewController, QrCodeReaderDelegate {
 }
+
+// MARK: - QRCodeDelegate
+extension QrViewController: QrCodeReaderDelegate {
```
out of curiosity: this is a style question to move delegate implementations to the extension, right?

nayooti replied:
Technically it makes no difference, so yes, it's a style question.

IMO it improves readability by separating a class after concerns. And if you use "MARK-statements" you can find them very easily. 

If you decide later to extract the delegate to a different class (for example if you want the class' viewModel to implement the delegate instead of the class), you can do so by basicly changing one word (`extension ViewModel: QRCodeReaderDelegate`).

cyBerta commented:
very cool that you implemented the ProgressAlertHandler this way so that every UIViewController can make use of it's functionality now.

I have to other things to comment: 
1. the string for qraccount_ask_create-and_login is still missing, but I think you are aware of that and there's probably a plan for it, right?
2. I tried an invalid QR code and what I got was this Alert Dialog. I don't know if this is intended.
![Screen Shot 2020-04-06 at 13 11 41](https://user-images.githubusercontent.com/2614900/78554352-48d34480-780b-11ea-91d3-1902cf163015.png)

nayooti commented:
@cyBerta As stated in Issue-specs the localised strings are still missing. 

I'll look into the Error-alert. I possibly mixed two alert-calls (calling the wrong alert function).
Thank you!

nayooti commented:
@cyBerta 
I checked the error-alert you mentioned. If you scan a non-dc-related QRcode you get this warning (translations are still missing):
![IMG_5EB752BAB5FB-1](https://user-images.githubusercontent.com/5249565/79209580-ea7a1780-7e43-11ea-813a-03f95384d6f5.jpeg)

I assume you scanned a dc-related qr code where the account creation failed. Did you?

Here it does show a plain-error alert as you showed in your screenshot. I stated in the PR-description above (in bold text), that we need a more specified error-message for this case.

cyBerta commented:
G

> I stated in the PR-description above (in bold text), that we need a more specified error-message for this case.

👍 Then let's create a string key for it and add that to the alert!
In general, we have the option to add yet untranslated strings to the untranslated.xml which you can find in deltachat-ios/tools/untranslated.xml . If you run` ./tools/tx-pull-translations.sh ` afterwards the untranslated strings will be shown on runtime (instead of the key).

nayooti commented:
@cyBerta Cool! Did just that.
